---
title: Libraries QC report
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output:
  html_document:
    theme: cosmo
    toc: true
    toc_depth: 2
    toc_float: true
    df_print: paged
    self_contained: true
params:
  rmd: libraries_qc.Rmd
  sample_id: null
  droplet_qc: null
  outdir: null
  features_matrix: null
  fragments: null
  adt_isotype: null
  gex_filters: !r list("library_size" = TRUE, "library_complexity" = TRUE, "mitochondrial_fraction" = TRUE)
  atac_filters: !r list("peak_count" = TRUE, "peak_fraction" = TRUE, "blacklist_fraction" = TRUE, "nucleosome_signal" = TRUE, "tss_enrichment" = TRUE)
  adt_filters: !r list("zero_ambient" = TRUE, "high_ambient" = TRUE, "high_controls" = TRUE)
---

```{r setup, include=FALSE}
# Set chunk options
knitr::opts_chunk$set(
  echo = FALSE,
  error = TRUE,
  message = FALSE,
  warning = FALSE,
  comment = "",
  results = "asis"
)

# Load packages
library(dplyr)
library(ggplot2)
library(plotly)
library(Matrix)
library(DropletUtils)
library(scuttle)
library(scater)
library(scran)
library(Seurat)
library(Signac)

# Check if running within Snakemake pipeline
if (exists("snakemake")) {
  # if running within Snakemake pipeline
  log <- file(snakemake@log[[1]], open = "w")
  sink(log, type = "message") # write stderr to Snakemake log file

  snakemake@source("utils.R")

  unlockBinding("params", .GlobalEnv)
  params$sample_id <- snakemake@wildcards$sample
  params <- update.params(snakemake@params)
  lockBinding("params", .GlobalEnv)

  if (!dir.exists(dirname(snakemake@output[[1]]))) dir.create(dirname(snakemake@output[[1]]), recursive = TRUE)
} else {
  # if running in isolation
  source("utils.R")
}

# Replace path placeholders
unlockBinding("params", .GlobalEnv)
for (x in c("droplet_qc", "outdir", "features_matrix", "fragments")) {
  params[[x]] <- glue::glue(params[[x]], sample = params$sample_id)
  if (length(params[[x]]) == 0) params[[x]] <- NULL # override default value of glue::glue() for null input (character(0)) and return null output
}
lockBinding("params", .GlobalEnv)

# Check for missing required params
required <- c(
  "sample_id",
  "droplet_qc",
  "outdir",
  "features_matrix"
)
is.missing <- sapply(params[required], is.null)
if (any(is.missing)) {
  names <- paste(names(params[required])[is.missing], collapse = ", ")
  stop(sprintf("The following required parameters are missing: %s", names))
}

# Initialise results list
bc <- list()

# Initialise list index
list.index <- 1

# Get feature modalities
modalities <- names(get.10x.h5(params$features_matrix))
```

### Sample: `r params$sample_id`

# Overview

```{r overview}
cat(prepend.index("Load count matrices and subset to common barcodes that passed droplet processing QC"), "\n")
if ("Gene Expression" %in% modalities) {
  cat(prepend.index("Calculate GEX QC metrics and discard outliers on 1 or more metrics"), "\n")
}
if ("Peaks" %in% modalities) {
  cat(prepend.index("Calculate ATAC QC metrics and discard outliers on 1 or more metrics"), "\n")
}
if ("Antibody Capture" %in% modalities) {
  cat(prepend.index("Calculate ADT QC metrics and discard outliers on 1 or more metrics"), "\n")
}
cat(prepend.index("Output text file containing list of barcodes passing all QC filters"), "\n")
```

# Data loading

```{r load-droplet-qc-results}
cat("* **Droplet processing QC results:**", file.path(params$droplet_qc, "metadata.tsv"), "\n")
droplet.qc.res <- vroom::vroom(
  file.path(params$droplet_qc, "metadata.tsv"),
  delim = "\t",
  col_names = TRUE,
  show_col_types = FALSE
) %>%
  mutate(Sample = factor(Sample, levels = sort(unique(Sample)), ordered = TRUE)) %>%
  as.data.frame()
rownames(droplet.qc.res) <- droplet.qc.res$Barcode
bc$cells <- rownames(droplet.qc.res)
```

```{r load-features-matrix, eval=!is.null(params$features_matrix)}
cat("* **Features matrix:**", params$features_matrix, "\n")
features.mat <- get.10x.h5(params$features_matrix)
bc$features <- colnames(features.mat[[1]])
```

```{r subset-barcodes}
bc.common <- Reduce(intersect, bc)
summary.stats <- data.frame(row.names = bc.common)
droplet.qc.res <- droplet.qc.res[bc.common, ]
features.mat <- lapply(features.mat, function(x) x[, bc.common])

# Re-initialise results list
bc <- list()
```

# Data processing

```{r gex-qc-metrics, eval=!is.null(features.mat[["Gene Expression"]])}
cat("## GEX QC metrics {.tabset}", "\n\n")
sce <- SingleCellExperiment(assays = list(counts = features.mat[["Gene Expression"]])) %>%
  add.cell.metadata(droplet.qc.res)
is.mito <- grep(pattern = "^MT-", ensembldb::mapIds(EnsDb.Hsapiens.v86::EnsDb.Hsapiens.v86, keys = rownames(sce), keytype = "GENEID", column = "SYMBOL")) %>%
  suppressWarnings() # pattern matching for mitochondrial genes
sce <- addPerCellQC(sce, subsets = list(mito = is.mito))
qc <- data.frame(row.names = colnames(sce))
if (params$gex_filters$library_size) qc$low_lib_size <- isOutlier(sce$sum, type = "lower", log = TRUE)
if (params$gex_filters$library_complexity) qc$low_lib_complexity <- isOutlier(sce$detected, type = "lower", log = TRUE)
if (params$gex_filters$mitochondrial_fraction) qc$high_mito_percent <- isOutlier(sce$subsets_mito_percent, type = "higher", log = FALSE)
qc$discard <- combine.filters(qc)
sce <- add.cell.metadata(sce, qc)
bc$gex <- colnames(sce[, !sce$discard])
summary.stats$GEX <- !qc$discard
```

```{r gex-qc-library-size, eval=all(!is.null(features.mat[["Gene Expression"]]), params$gex_filters$library_size)}
cat("### Library size", "\n\n")
ggplotly(
  violin.plot(
    colData(sce),
    x = Sample,
    y = sum,
    colour = low_lib_size,
    size = 0.1,
    title = "Low library size",
    x.lab = "Sample",
    y.lab = "Total UMI count",
    log = TRUE
  ) +
    scale_colour_manual(values = c("grey80", "#d82526")),
  tooltip = c("y", "colour")
)
```

```{r gex-qc-library-complexity, eval=all(!is.null(features.mat[["Gene Expression"]]), params$gex_filters$library_complexity)}
cat("### Library complexity", "\n\n")
ggplotly(
  violin.plot(
    colData(sce),
    x = Sample,
    y = detected,
    colour = low_lib_complexity,
    size = 0.1,
    title = "Low library complexity",
    x.lab = "Sample",
    y.lab = "Total genes detected",
    log = TRUE
  ) +
    scale_colour_manual(values = c("grey80", "#d82526")),
  tooltip = c("y", "colour")
)
```

```{r gex-qc-mitochondrial-fraction, eval=all(!is.null(features.mat[["Gene Expression"]]), params$gex_filters$mitochondrial_fraction)}
cat("### Mitochondrial fraction", "\n\n")
ggplotly(
  violin.plot(
    colData(sce),
    x = Sample,
    y = subsets_mito_percent,
    colour = high_mito_percent,
    size = 0.1,
    title = "High mitochondrial fraction",
    x.lab = "Sample",
    y.lab = "Percentage mitochondrial genes"
  ) +
    scale_colour_manual(values = c("grey80", "#d82526")),
  tooltip = c("y", "colour")
)
```

```{r gex-qc-diagnostics, eval=!is.null(features.mat[["Gene Expression"]])}
cat("### Diagnostics", "\n\n")
cat("* **Cells passed:**", length(bc$gex) %>% format(big.mark = ","), "\n")
cat("* **Cells excluded:**", sum(sce$discard, na.rm = TRUE) %>% format(big.mark = ","), "\n")
if (params$gex_filters$library_size) {
  cat("  + **Low library size:**", sum(sce$low_lib_size, na.rm = TRUE) %>% format(big.mark = ","), "\n")
}
if (params$gex_filters$library_complexity) {
  cat("  + **Low library complexity:**", sum(sce$low_lib_complexity, na.rm = TRUE) %>% format(big.mark = ","), "\n")
}
if (params$gex_filters$mitochondrial_fraction) {
  cat("  + **High mitochondrial fraction:**", sum(sce$high_mito_percent, na.rm = TRUE) %>% format(big.mark = ","), "\n")
}
cat("\n\n")
scatter.plot(
  colData(sce) %>%
    as.data.frame() %>%
    arrange(discard),
  x = sum,
  y = subsets_mito_percent,
  size = 0.1,
  colour = discard,
  title = "Cells excluded (library size vs mitochondrial fraction)",
  x.lab = "Total UMI count",
  y.lab = "Percentage mitochondrial genes",
  log = "x"
) +
  scale_colour_manual(values = c("grey80", "#d82526"))
cat("\n\n")
scatter.plot(
  colData(sce) %>%
    as.data.frame() %>%
    arrange(discard),
  x = sum,
  y = detected,
  size = 0.1,
  colour = discard,
  title = "Cells excluded (library size vs library complexity)",
  x.lab = "Total UMI count",
  y.lab = "Total genes detected",
  log = TRUE
) +
  scale_colour_manual(values = c("grey80", "#d82526"))
```

```{r atac-qc-metrics, eval=!is.null(features.mat[["Peaks"]])}
cat("## ATAC QC metrics {.tabset}", "\n\n")
fragments <- CreateFragmentObject(params$fragments, cells = bc.common)
annotation <- GetGRangesFromEnsDb(EnsDb.Hsapiens.v86::EnsDb.Hsapiens.v86) %>% suppressWarnings()
seqlevelsStyle(annotation) <- "UCSC"
seu <- CreateSeuratObject(
  counts = CreateChromatinAssay(
    counts = features.mat[["Peaks"]],
    sep = c(":", "-"),
    fragments = fragments,
    annotation = annotation
  ),
  assay = "ATAC"
) %>%
  add.cell.metadata(droplet.qc.res)
qc <- data.frame(row.names = colnames(seu))
if (params$atac_filters$peak_count) {
  seu$sum <- colSums(seu)
  qc$low_peak_count <- isOutlier(seu$sum, type = "lower", log = TRUE)
}
if (params$atac_filters$peak_fraction) {
  fragments.count <- CountFragments(fragments@path)
  rownames(fragments.count) <- fragments.count$CB
  seu$fragments <- fragments.count[colnames(seu), "frequency_count"]
  seu <- FRiP(seu, assay = "ATAC", total.fragments = "fragments")
  if (any(seu$FRiP > 1)) seu$FRiP <- seu$FRiP / 2 # if any FRiP > 1, assume peak counts represent transposition events rather than fragments and divide by 2 to adjust
  qc$low_peak_fraction <- seu$FRiP < 0.2 # hard threshold of 0.2 as per ENCODE consortium guidelines
}
if (params$atac_filters$blacklist_fraction) {
  seu$blacklist_fraction <- FractionCountsInRegion(seu, regions = blacklist_hg38_unified)
  qc$high_blacklist_fraction <- isOutlier(seu$blacklist_fraction, type = "higher", log = FALSE, nmads = 8) # only exclude extreme outliers
}
if (params$atac_filters$nucleosome_signal) {
  seu <- NucleosomeSignal(seu)
  qc$high_nucleosome_signal <- isOutlier(seu$nucleosome_signal, type = "higher", log = FALSE, nmads = 8) # only exclude extreme outliers
}
if (params$atac_filters$tss_enrichment) {
  seu <- TSSEnrichment(seu, fast = FALSE)
  qc$low_tss_enrichment <- isOutlier(seu$TSS.enrichment, type = "lower", log = FALSE)
}
qc$discard <- combine.filters(qc)
seu <- add.cell.metadata(seu, qc)
bc$atac <- colnames(seu[, !seu$discard])
summary.stats$ATAC <- !qc$discard
```

```{r atac-qc-peak-count, eval=all(!is.null(features.mat[["Peaks"]]), params$atac_filters$peak_count)}
cat("### Peak count", "\n\n")
ggplotly(
  violin.plot(
    seu[[]],
    x = Sample,
    y = sum,
    colour = low_peak_count,
    size = 0.1,
    title = "Low total reads in peaks",
    x.lab = "Sample",
    y.lab = "Total reads in peaks",
    log = TRUE
  ) +
    scale_colour_manual(values = c("grey80", "#d82526")),
  tooltip = c("y", "colour")
)
```

```{r atac-qc-peak-fraction, eval=all(!is.null(features.mat[["Peaks"]]), params$atac_filters$peak_fraction)}
cat("### Peak fraction", "\n\n")
ggplotly(
  violin.plot(
    seu[[]],
    x = Sample,
    y = FRiP * 100,
    colour = low_peak_fraction,
    size = 0.1,
    title = "Low fraction reads in peaks",
    x.lab = "Sample",
    y.lab = "Percentage reads in peaks"
  ) +
    scale_colour_manual(values = c("grey80", "#d82526")),
  tooltip = c("y", "colour")
)
```

```{r atac-qc-blacklist-fraction, eval=all(!is.null(features.mat[["Peaks"]]), params$atac_filters$blacklist_fraction)}
cat("### Blacklist fraction", "\n\n")
ggplotly(
  violin.plot(
    seu[[]],
    x = Sample,
    y = blacklist_fraction * 100,
    colour = high_blacklist_fraction,
    size = 0.1,
    title = "High blacklist fraction",
    x.lab = "Sample",
    y.lab = "Percentage reads in genomic blacklist regions"
  ) +
    scale_colour_manual(values = c("grey80", "#d82526")),
  tooltip = c("y", "colour")
)
```

```{r atac-qc-nucleosome-signal, eval=all(!is.null(features.mat[["Peaks"]]), params$atac_filters$nucleosome_signal)}
cat("### Nucleosome signal", "\n\n")
ggplotly(
  violin.plot(
    seu[[]],
    x = Sample,
    y = nucleosome_signal,
    colour = high_nucleosome_signal,
    size = 0.1,
    title = "High nucleosome signal",
    x.lab = "Sample",
    y.lab = "Nucleosome signal"
  ) +
    scale_colour_manual(values = c("grey80", "#d82526")),
  tooltip = c("y", "colour")
)
```

```{r atac-qc-tss-enrichment, eval=all(!is.null(features.mat[["Peaks"]]), params$atac_filters$tss_enrichment)}
cat("### TSS enrichment", "\n\n")
ggplotly(
  violin.plot(
    seu[[]],
    x = Sample,
    y = TSS.enrichment,
    colour = low_tss_enrichment,
    size = 0.1,
    title = "Low TSS enrichment score",
    x.lab = "Sample",
    y.lab = "TSS enrichment score"
  ) +
    scale_colour_manual(values = c("grey80", "#d82526")),
  tooltip = c("y", "colour")
)
```

```{r atac-qc-diagnostics, fig.width=10, fig.height=5, eval=!is.null(features.mat[["Peaks"]])}
cat("### Diagnostics", "\n\n")
cat("* **Cells passed:**", length(bc$atac) %>% format(big.mark = ","), "\n")
cat("* **Cells excluded:**", sum(seu$discard, na.rm = TRUE) %>% format(big.mark = ","), "\n")
if (params$atac_filters$peak_count) {
  cat("  + **Low total reads in peaks:**", sum(seu$low_peak_count, na.rm = TRUE) %>% format(big.mark = ","), "\n")
}
if (params$atac_filters$peak_fraction) {
  cat("  + **Low fraction reads in peaks:**", sum(seu$low_peak_fraction, na.rm = TRUE) %>% format(big.mark = ","), "\n")
}
if (params$atac_filters$blacklist_fraction) {
  cat("  + **High blacklist region reads:**", sum(seu$high_blacklist_fraction, na.rm = TRUE) %>% format(big.mark = ","), "\n")
}
if (params$atac_filters$nucleosome_signal) {
  cat("  + **High nucleosome signal:**", sum(seu$high_nucleosome_signal, na.rm = TRUE) %>% format(big.mark = ","), "\n")
}
if (params$atac_filters$tss_enrichment) {
  cat("  + **Low TSS enrichment score:**", sum(seu$low_tss_enrichment, na.rm = TRUE) %>% format(big.mark = ","), "\n")
}
cat("\n\n")
FragmentHistogram(seu, group.by = "high_nucleosome_signal") +
  labs(title = "Nucleosome banding") +
  scale_fill_manual(values = c("grey80", "#d82526")) +
  theme_minimal() +
  theme(
    legend.position = "none",
    strip.text = element_blank()
  )
cat("\n\n")
TSSPlot(seu, group.by = "low_tss_enrichment") +
  scale_colour_manual(values = c("grey80", "#d82526")) +
  theme_minimal() +
  theme(
    legend.position = "none",
    strip.text = element_blank()
  )
```

```{r adt-qc-metrics, eval=!is.null(features.mat[["Antibody Capture"]])}
cat("## ADT QC metrics {.tabset}", "\n\n")
sce <- SingleCellExperiment(assays = list(counts = features.mat[["Antibody Capture"]])) %>%
  add.cell.metadata(droplet.qc.res)
filters <- NULL
if (is.null(params$adt_isotype)) {
  qc <- cleanTagCounts(sce)
  if (params$adt_filters$high_ambient) filters <- "high.ambient"
} else {
  is.control <- grep(pattern = params$adt_isotype, rownames(sce))
  qc <- cleanTagCounts(sce, controls = is.control)
  if (params$adt_filters$high_controls) filters <- "high.controls"
}
if (params$adt_filters$zero_ambient) filters <- c(filters, "zero.ambient")
qc$discard <- if (!is.null(filters)) combine.filters(lapply(filters, function(x) qc[[x]])) else FALSE
sce <- add.cell.metadata(sce, qc)
bc$adt <- colnames(sce[, !sce$discard])
summary.stats$ADT <- !qc$discard
```

```{r adt-ambient-contamination-isotype-control, eval=!is.null(features.mat[["Antibody Capture"]])}
if (is.null(params$adt_isotype)) {
  cat("### Ambient contamination", "\n\n")
  ggplotly(
    violin.plot(
      colData(sce),
      x = Sample,
      y = ambient.scale,
      colour = high.ambient,
      size = 0.1,
      title = "High ambient contamination",
      x.lab = "Sample",
      y.lab = "Relative ambient contamination"
    ) +
      scale_colour_manual(values = c("grey80", "#d82526")),
    tooltip = c("y", "colour")
  )
} else {
  cat("### Isotype control", "\n\n")
  ggplotly(
    violin.plot(
      colData(sce),
      x = Sample,
      y = sum.controls,
      colour = high.controls,
      size = 0.1,
      title = "High non-specific binding",
      x.lab = "Sample",
      y.lab = "Total isotype control antibody count"
    ) +
      scale_colour_manual(values = c("grey80", "#d82526")),
    tooltip = c("y", "colour")
  )
}
```

```{r adt-qc-diagnostics, eval=!is.null(features.mat[["Antibody Capture"]])}
cat("### Diagnostics", "\n\n")
cat("* **Cells passed:**", length(bc$adt) %>% format(big.mark = ","), "\n")
cat("* **Cells excluded:**", sum(sce$discard, na.rm = TRUE) %>% format(big.mark = ","), "\n")
if (params$adt_filters$zero_ambient) {
  cat("  + **Zero ambient contamination:**", sum(sce$zero.ambient, na.rm = TRUE) %>% format(big.mark = ","), "\n")
}
if (is.null(params$adt_isotype) && params$adt_filters$high_ambient) {
  cat("  + **High ambient contamination:**", sum(sce$high.ambient, na.rm = TRUE) %>% format(big.mark = ","), "\n")
}
if (!is.null(params$adt_isotype) && params$adt_filters$high_controls) {
  cat("  + **High non-specific binding:**", sum(sce$high.controls, na.rm = TRUE) %>% format(big.mark = ","), "\n")
}
```

# Summary

```{r output}
bc.pass <- Reduce(intersect, bc, init = rownames(droplet.qc.res))
summary.stats <- summary.stats %>%
  mutate(All = if_all(everything())) %>%
  tibble::rownames_to_column("Barcode")

if (!dir.exists(params$outdir)) dir.create(params$outdir, recursive = TRUE)
vroom::vroom_write_lines(
  bc.pass,
  file = file.path(params$outdir, "cell_barcodes.txt")
)
vroom::vroom_write(
  summary.stats,
  file = file.path(params$outdir, "qc_filters.tsv"),
  delim = "\t"
)

cat("* **Total cells passed:**", length(bc.pass) %>% format(big.mark = ","), "\n")
cat("* **Total cells excluded:**", (nrow(droplet.qc.res) - length(bc.pass)) %>% format(big.mark = ","), "\n")
if (!is.null(features.mat[["Gene Expression"]])) {
  cat("  + **Failed GEX QC:**", (nrow(droplet.qc.res) - length(bc$gex)) %>% format(big.mark = ","), "\n")
}
if (!is.null(features.mat[["Peaks"]])) {
  cat("  + **Failed ATAC QC:**", (nrow(droplet.qc.res) - length(bc$atac)) %>% format(big.mark = ","), "\n")
}
if (!is.null(features.mat[["Antibody Capture"]])) {
  cat("  + **Failed ADT QC:**", (nrow(droplet.qc.res) - length(bc$adt)) %>% format(big.mark = ","), "\n")
}
```

# Session Info

```{r session-info, results="markup"}
sessionInfo()
```

# Source
This report was automatically generated from an R Markdown script (<a download="libraries_qc.Rmd" href="`r base64enc::dataURI(file = params$rmd, mime = 'text/rmd', encoding = 'base64')`">download source</a>).
