---
title: "Batch correction report"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output:
  html_document:
    theme: "flatly"
    highlight: "haddock"
    toc: true
    toc_depth: 2
    toc_float: true
    df_print: "paged"
    self_contained: true
params:
  rmd: "batch_correction.Rmd"
  threads: 1
  sample_id: null
  droplet_qc: null
  libraries_qc: null
  output_dir: null
  gex_matrix: null
  adt_matrix: null
  adt_filters: !r list("isotype_control" = NULL)
  control: null
  correction_method:
    - "none"
    - "rescale"
    - "fastmnn"
  correction_projection: "TSNE"
---

```{r setup, include=FALSE}
# Set chunk options
knitr::opts_chunk$set(
  echo = FALSE,
  error = TRUE,
  message = FALSE,
  warning = FALSE,
  comment = "",
  results = "asis"
)

# Load packages
library(future.apply)
library(dplyr)
library(ggplot2)
library(plotly)
library(scuttle)
library(scater)
library(scran)
library(batchelor)

# Initialise plotly dependencies
ggplotly(ggplot())

# Check if running within Snakemake pipeline
if (exists("snakemake")) {
  # if running within Snakemake pipeline
  log <- file(snakemake@log[[1]], open = "w")
  sink(log, type = "message") # write stderr to Snakemake log file

  snakemake@source("functions.R")

  unlockBinding("params", .GlobalEnv)
  params$threads <- snakemake@threads
  params <- update.params(snakemake@params)
  lockBinding("params", .GlobalEnv)

  if (!dir.exists(dirname(snakemake@output[[1]]))) dir.create(dirname(snakemake@output[[1]]), recursive = TRUE)
} else {
  # if running in isolation
  source("functions.R")
}

# Replace path placeholders
unlockBinding("params", .GlobalEnv)
for (x in c("droplet_qc", "libraries_qc", "gex_matrix", "adt_matrix")) {
  params[[x]] <- glue::glue(params[[x]], sample = params$sample_id)
  if (length(params[[x]]) == 0) params[[x]] <- NULL # override default value of glue::glue() for null input (character(0)) and return null output
}
lockBinding("params", .GlobalEnv)

# Check for missing required params
required <- c(
  "sample_id",
  "droplet_qc",
  "libraries_qc",
  "output_dir"
)
is.missing <- sapply(params[required], is.null)
if (any(is.missing)) {
  names <- paste(names(params[required])[is.missing], collapse = ", ")
  stop(sprintf("The following required parameters are missing: %s", names))
}

# Set parallel processing limits
plan(multicore, workers = params$threads)
options(future.globals.maxSize = 1000 * 1024^2) # set default global variable size limit to 1GB

# Parse correction method
batchelor.param <- lapply(
  params$correction_method,
  function(x) {
    if (x == "none") {
      return(NoCorrectParam())
    } else if (x == "rescale") {
      return(RescaleParam())
    } else if (x == "fastmnn") {
      return(FastMnnParam(cos.norm = FALSE)) # cosine normalisation NOT required as using batchelor::multiBatchNorm() to normalise counts for differences in sequencing depth across batches
    } else {
      stop(sprintf("Invalid correction method specified: '%s'", x), "; valid options are 'none', 'rescale' and 'fastmnn'")
    }
  }
) %>%
  setNames(params$correction_method)

# Initialise results list
sce <- list()
```

```{r samples}
cat("### Samples:", "\n\n")
cat("```", paste(params$sample_id, collapse = "\n"), "```", sep = "\n")
```

# Overview

1. Load count matrices and subset to barcodes that passed libraries processing QC
2. Perform batch correction for each library type using specified method(s).
If a control sample (i.e. cross-batch technical replicate) is specified, batch
correction is computed using only cells from the control sample in each batch
then extrapolated to all other cells.
3. Output RDS files containing batch-corrected `SingleCellExperiment` objects for each
library type and each batch correction method. If a control sample is specified, cells
from the control sample in each batch will be discarded from the output.

# Data loading

```{r load-libraries-qc-results}
cat("* **Libraries QC results:**", "\n\n")
cat("```", paste(params$libraries_qc, collapse = "\n"), "```", sep = "\n")
cells <- future_lapply(
  params$libraries_qc,
  function(file) {
    x <- vroom::vroom(
      file,
      delim = ",",
      col_names = FALSE,
      show_col_types = FALSE
    ) %>%
      pull(1)
  }
) %>%
  setNames(params$sample_id)
```

```{r load-droplet-qc-results}
cat("* **Droplet processing QC results:**", "\n\n")
cat("```", paste(params$droplet_qc, collapse = "\n"), "```", sep = "\n")
droplet.qc.res <- future_Map(
  function(file, batch, cells) {
    df <- vroom::vroom(
      file,
      delim = ",",
      col_names = TRUE,
      show_col_types = FALSE
    ) %>%
      mutate(Sample = factor(Sample)) %>%
      as.data.frame()
    df <- df[match(cells, df$Barcode), ]
    rownames(df) <- paste(batch, df$Barcode, sep = "_")
    return(df)
  },
  params$droplet_qc,
  params$sample_id,
  cells
) %>%
  setNames(params$sample_id)
restrict <- lapply(
  droplet.qc.res,
  function(x) {
    if (!is.null(params$control)) {
      x$Sample == params$control
    } else {
      NULL
    }
  }
) %>%
  setNames(params$sample_id)
```

```{r load-gex-matrix, eval=!is.null(params$gex_matrix)}
cat("* **GEX matrices:**", "\n\n")
cat("```", paste(params$gex_matrix, collapse = "\n"), "```", sep = "\n")
gex.mat <- future_Map(
  function(file, batch, cells) {
    mat <- get.10x.matrix(
      file,
      cells = cells,
      type = "Gene Expression"
    )
    colnames(mat) <- paste(batch, colnames(mat), sep = "_")
    return(mat)
  },
  params$gex_matrix,
  params$sample_id,
  cells
) %>%
  setNames(params$sample_id)
```

```{r load-adt-matrix, eval=!is.null(params$adt_matrix)}
cat("* **ADT matrices:**", "\n\n")
cat("```", paste(params$adt_matrix, collapse = "\n"), "```", sep = "\n")
adt.mat <- future_Map(
  function(adt_file, gex_file, batch, cells) {
    if (adt_file == gex_file) {
      mat <- get.10x.matrix(
        adt_file,
        cells = cells,
        type = "Antibody Capture"
      )
      mat <- mat[grep(pattern = "^ADT_", rownames(mat)), ]
    } else {
      mat <- get.barcounter.matrix(
        adt_file,
        cells = cells,
        features.pattern = "^ADT_",
        add.suffix = TRUE
      )
    }
    colnames(mat) <- paste(batch, colnames(mat), sep = "_")
    if (!is.null(params$adt_filters$isotype_control)) {
      mat <- mat[grep(pattern = params$adt_filters$isotype_control, rownames(mat), invert = TRUE), ] # remove isotype control antibodies
    }
    return(mat)
  },
  params$adt_matrix,
  params$gex_matrix,
  params$sample_id,
  cells
) %>%
  setNames(params$sample_id)
```

# Data processing

```{r gex-batch-correction, eval=!is.null(params$gex_matrix)}
cat("## GEX batch correction {.tabset}", "\n\n")
sce$gex <- future_Map(
  function(matrix, metadata) {
    SingleCellExperiment(
      assays = list(counts = matrix)
    ) %>%
      add.cell.metadata(metadata) %>%
      computePooledFactors(clusters = quickCluster(.))
  },
  gex.mat,
  droplet.qc.res,
  future.seed = 100
) %>%
  multiBatchNorm()
sce$gex <- future_lapply(
  batchelor.param,
  function(correction.method) {
    correctExperiments(
      sce$gex,
      restrict = restrict,
      PARAM = correction.method
    )
  },
  future.seed = 100
) %>%
  setNames(params$correction_method)

for (x in names(sce$gex)) {
  correction.method <- case_when(
    x == "none" ~ "No correction",
    x == "rescale" ~ "Rescale",
    x == "fastmnn" ~ "Fast MNN"
  )
  cat("###", correction.method, "\n\n")
  ggplotly(
    batch.plot(
      sce$gex[[x]],
      control = params$control,
      projection = params$correction_projection,
      size = 1,
      stroke = 0.2,
      random.seed = 100,
      title = correction.method
    ) +
      scale_colour_manual(values = c(NA, "#d82526")),
    tooltip = c("colour", "fill")
  ) %>%
    htmltools::tagList() %>%
    print()
  cat("\n\n")
}
```


```{r adt-batch-correction, eval=!is.null(params$adt_matrix)}
cat("## ADT batch correction {.tabset}", "\n\n")
sce$adt <- future_Map(
  function(matrix, metadata) {
    SingleCellExperiment(
      assays = list(counts = matrix)
    ) %>%
      add.cell.metadata(metadata) %>%
      computeGeometricFactors()
  },
  adt.mat,
  droplet.qc.res
) %>%
  multiBatchNorm()
sce$adt <- future_lapply(
  batchelor.param,
  function(correction.method) {
    correctExperiments(
      sce$adt,
      restrict = restrict,
      PARAM = correction.method
    )
  },
  future.seed = 100
) %>%
  setNames(params$correction_method)

for (x in names(sce$adt)) {
  correction.method <- case_when(
    x == "none" ~ "No correction",
    x == "rescale" ~ "Rescale",
    x == "fastmnn" ~ "Fast MNN"
  )
  cat("###", correction.method, "\n\n")
  ggplotly(
    batch.plot(
      sce$adt[[x]],
      control = params$control,
      projection = params$correction_projection,
      size = 1,
      stroke = 0.2,
      random.seed = 100,
      title = correction.method
    ) +
      scale_colour_manual(values = c(NA, "#d82526")),
    tooltip = c("colour", "fill")
  ) %>%
    htmltools::tagList() %>%
    print()
  cat("\n\n")
}
```

# Output

```{r output}
if (!dir.exists(params$output_dir)) dir.create(params$output_dir, recursive = TRUE)
cat("```", "\n")
for (x in names(sce)) {
  for (y in names(sce[[x]]) %>% setdiff("none")) {
    if (!is.null(params$control)) {
      control <- sce[[x]][[y]]$Sample == params$control
      sce[[x]][[y]] <- sce[[x]][[y]][, !control]
    }
    saveRDS(
      sce[[x]][[y]],
      file = file.path(params$output_dir, sprintf("%s_%s.rds", x, y))
    )
    cat(sprintf("%s_%s.rds", x, y), "\n")
    print(sce[[x]][[y]])
    cat("\n")
  }
}
cat("```", "\n")
```

# Session Info

```{r session-info, results="markup"}
sessionInfo()
```

# Source
This report was automatically generated from an R Markdown script (<a download="batch_correction.Rmd" href="`r base64enc::dataURI(file = params$rmd, mime = 'text/rmd', encoding = 'base64')`">download source</a>).
